articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,235 @@
+
+The new QoS policy structs will have additional fields that use enums based on the values defined above to specify the desired QoS settings.
+Each enum field instance in the struct will also have an associated number field added to the struct that represents a time value for the policy.
+
+#### Assert Liveliness Functions
+
+New functions will need to be added that can be used by the application to explicitly assert liveliness.
+One function to assert liveliness at the Node level, one to assert it at the Topic level, and one to assert it at the Service level.
+While liveliness will also be implicitly assumed just based on sending messages, these functions will be used by the application to explicitly declare resources are alive.
+These functions will need to be implemented in every layer from the rmw up through the rcl and language specific client libraries, such as rclcpp.
+
+#### rcl_wait and rmw_wait
+
+The rcl layer is currently using a WaitSet in order to be informed of events from the rmw layer, such as incoming messages.
+These WaitSets contain lists of several types of conditions, such as timers firing or subscriptions receiving data.
+In order to support new Topic Status Events, a new type will be added to the existing WaitSet and the rmw layer will set them when these events occur.
+
+#### rcl_take_status and rmw_take_status
+
+New functions called rcl_take_status and rmw_take_status will need to be added that can directly query the status for a Topic/Service.
+It will operate in a similar manner to the rcl_take and rmw_take functions that are used to retrieve messages for a subscription.
+It will be used by the executors when they receive a notice via the waitset mentioned above that a resource has a new status event available.
+
+## RMW Vendor Support
+
+All of these new QoS policies will need to be supported in the rmw implementations.
+As we add new QoS policies it is likely that not all rmw vendors will provide support for every QoS policy that ROS defines.
+This is especially true for non-DDS based implementations that do not already have native support for these features.
+Because of this we need a way for the application to know if the rmw vendor that is being used supports the specified policies at the specified levels.
+
+In the case where an rmw vendor does not support a requested QoS they can do one of two things.
+They could fail with an exception that notes that an unsupported QoS policy was requested or they can choose to continue operating with a reduced QoS policy.
+In the case where a vendor chooses to continue operating with a reduced policy it must trigger a status event for the resource that can be handled by the new callback outlined above.
+This provides an application a way of being notified that even though their resource is operating, it is doing so with a reduced QoS.
```

Comment:
I find this second option dangerous. Experience shows that if users have the option they will use and they will ignore warnings on the status events, out of ignorance or out of not understanding the limitation. I'd vote for raising an exception as being the only option.

Reply:
Raising an exception (or returning a non-SUCCESS error code as the case may be) then requires every rmw implementation to implement every QoS policy or risk nothing being able to run. Yes, you now have the risk of users ignoring the warnings, but it also allows the application to actually run in cases where you don't actually care. 

For example, I imagine that a large number of commonly used Nodes are going to provide LIVELINESS_MANUAL_TOPIC to allow the most verbose tracking of liveliness by consumers even though it isn't really required for them to run. There may be an rmw implementation that a user wants to use for their particular use case that does not implement the liveliness QoS policy. The user may have decided that they don't actually need liveliness tracked in their system but do need some other feature offered by this particular rmw implementation. In that scenario, throwing an exception means that the user will not be able to use any of those common Nodes that offer that offer liveliness because the thrown exception would prevent them from starting up. On the other hand, going the route of only making this a warning allows the Nodes to still run without the use of liveliness and the user can still use them. 

I guess the tradeoff here is that by not using exceptions  we allow Nodes to be more portable out of the box when the implementor doesn't fully handle the situation when an rmw implementation doesn't support a QoS policy. However, it does push some burden onto the end user of the Node to understand the impact of their rmw choice.

Reply:
> Raising an exception (or returning a non-SUCCESS error code as the case may be) then requires every rmw implementation to implement every QoS policy or risk nothing being able to run. Yes, you now have the risk of users ignoring the warnings, but it also allows the application to actually run in cases where you don't actually care.

If the feature is not implemented uniformly across middleware implementations, then I question it's value in the first place. Making the API optional just makes user code more complicated.

In the case of emitting warnings, I think that's simply not good enough. If I write a program and expect the deadline (for example) to be respected, but it fails to do so, only emitting a warning, then my program may no longer be valid. Instead, I'd like the API to throw (or return an error code), and if you want to write code that can use it optionally, then add an API call which can be used to check and see if it is not implemented, in which case the user can avoid calling it, or configure their system differently. Or they can just call it and check for the error state, catch it / handle it and continue with an alternative code path that doesn't require it.

In general, however, I would prefer to avoid optional components of the API due to the complexity they add at all layers of the system. Instead we should aim to only include features in the API that are useful enough to require all middleware implementations to support one way or the other. This contributes to my personal preference of including as few QoS settings as we can manage and still do useful stuff. The breadth of configuration is useful, but makes portability difficult and implementations complicated. Having optional API's try's to address this, but fails to avoid the complexity in my opinion.